feat: create cc_toolchains for multiple exec platforms

BrandonThomasJonesARM opened the pull request:
This change aims to allow hermetic_cc_toolchain to create toolchains for multiple exec platforms. It now generates repositories of the format {exec}-zig_sdk and registers each toolchain to only be exec_compatible_with those specific platform constraints.

I want this PR to be a WIP as we discuss how we want this to look and if you're happy with it's general flow.

This is an attempt at fixing https://github.com/uber/hermetic_cc_toolchain/issues/148 

So far I have tested that my local client (mac) can build c code using the toolchains registered on a `x86_64` remote executor and have seen success. It'll will be hard for me to test every combination; but I will be able to local (x86_64) -> remote (arm64) tests as well.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/uber/hermetic_cc_toolchain?pullRequest=196) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/uber/hermetic_cc_toolchain?pullRequest=196) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/uber/hermetic_cc_toolchain?pullRequest=196) it.</sub>

BrandonThomasJonesARM commented:
@linzhp @motiejus Apologies, not sure who is best to ping, but some early eyes on this would be great to understand if you're happy with it's direction.

BrandonThomasJonesARM commented on toolchain/platform/defs.bzl:
```diff
@@ -28,7 +28,7 @@ def declare_libc_aware_platforms():
                 suffix = "_{}".format(libc),
-                extra_constraints = ["@zig_sdk//libc:{}".format(libc)],
+                extra_constraints = ["//libc:{}".format(libc)],
             )
```
The reason we can do this is because these starlark functions are always called from the generated repository, we do not need to use an external repository label here. We stay local to the {exec}-zig_sdk repository in these cases.

BrandonThomasJonesARM commented on toolchain/defs.bzl:
```diff
@@ -213,12 +234,21 @@ def _zig_repository_impl(repository_ctx):
 
+    urls = [uf.format(**format_vars_exec) for uf in repository_ctx.attr.url_formats]
+
+    repository_ctx.download_and_extract(
```
Something to discuss since you will understand the flow better than me; this is my "hack" for getting the exec toolchain in place since we need the host toolchain to build the zig-wrapper.

I believe this is fine as zig is small enough; however it isn't ideal. I need to think about if there is a way to build the `zig-wrapper` executable outside of the repository_ctx or not; otherwise maybe having a `@host-zig_sdk` could work, but I worry about how to keep the cache hit rate good if we're using non-relative paths + we'd have to ensure that repository is local, which I do think the `local` parameter will handle for us, but I need to think more on it.

linzhp replied:
yeah, I don't see a way to avoid downloading two Zig SDKs when host and exec platforms are different. Building zig-wrapper in repository rules is an intentional design to read `HERMETIC_CC_TOOLCHAIN_CACHE_PREFIX`, which is not possible during regular build actions.

linzhp commented on toolchain/defs.bzl:
```diff
@@ -78,29 +80,40 @@ def toolchains(
 
     zig_repository(
-        name = "zig_sdk",
+        name = "{}-zig_sdk".format(exec),
```
```suggestion
        name = "zig_sdk-{}".format(exec),
```

Nit: let's put the platform name after zig_sdk

linzhp commented on toolchain/ext.bzl:
```diff
@@ -1,6 +1,19 @@
 load("@hermetic_cc_toolchain//toolchain:defs.bzl", zig_toolchains = "toolchains")
 
+_HOSTS = (
```
Let's declare a list of tuples like rules_go does, so we don't need to split it later: 

https://github.com/bazel-contrib/rules_go/blob/87e66bb70858570ac668f95662d6070417d5a416/go/private/extensions.bzl#L99-L106

BrandonThomasJonesARM commented:
Thanks for the initial look @linzhp 

I have addressed the comments made, I have added a `local` repository rule for `zig_sdk` that symlinks to the exec repo that matches the host. This is to try and ensure backwards compatibility somewhat.  I am not entirely confident with it; it might cause more problems than it solves.

linzhp commented on rules/zig_binary.bzl:
```diff
@@ -32,11 +32,11 @@ zig_binary = rule(
         ),
         "_zig": attr.label(
-            default = "@zig_sdk//:tools/zig-wrapper",
+            default = "@zig_sdk-linux-amd64//:tools/zig-wrapper",
```
Does this assume the execution platform to be always linux amd64?

BrandonThomasJonesARM replied:
Yes, this was down for testing before I added the patch that creates the `zig_sdk` that auto-resolves which one is host for you, so will fix this up.

BrandonThomasJonesARM replied:
If these are only used for local tests and not with RE, we can likely leave this be for now.

linzhp commented on toolchain/defs.bzl:
```diff
@@ -78,29 +80,44 @@ def toolchains(
-        os = "windows"
-
-    host_platform = "{}-{}".format(os, arch)
+    arch_os = repository_ctx.attr.exec.split("-")
+    exec_os = arch_os[0]
+    exec_arch = arch_os[1]
```
Let's pass the exec_os and exec_arch as attributes, so we don't need to split here.
```suggestion
    exec_os = repository_ctx.attr.exec_os
    exec_arch = repository_ctx.attr.exec_arch
```

linzhp commented on toolchain/defs.bzl:
```diff
@@ -230,11 +264,26 @@ zig_repository = repository_rule(
+    repository_ctx.symlink(str(compatible_zig_path) + "/..", ".")
+
+
+local_zig_repository = repository_rule(
```
```suggestion
host_zig_repository = repository_rule(
```
To avoid confusing with Bazel [local repository rules](https://bazel.build/rules/lib/repo/local).

BrandonThomasJonesARM replied:
Done

BrandonThomasJonesARM commented:
I have tested the latest iteration with RE where host was x86 and my remote was arm64, and got successful compilations. I cannot run the full test suite yet as i'm running into a problem with `rules_go` that I am looking into.

BrandonThomasJonesARM commented:
@linzhp ping/bump , is there anything else outstanding on this PR you currently want me to address?

linzhp commented:
I can take a look sometime this week

linzhp commented:
There is some issues with Windows' platform names though

BrandonThomasJonesARM commented:
> There is some issues with Windows' platform names though

Let me investigate

BrandonThomasJonesARM commented:
Pushed a change that hopefully fixed the failing worksflows, could we try again please? Is there any way for me to run the majority of it locally outside of `bazel test //...` to ensure I am not wasting unnecessary resource time testing it in your CI flow?

BrandonThomasJonesARM commented:
Looks like macos fails for reasons outside my control; see https://github.com/uber/hermetic_cc_toolchain/issues/131

Is the testing for macos cross-compiling go meant to be work?

BrandonThomasJonesARM commented:
Fixed all failures outside of the releaase checks, I have attempted to change the API for the WORKSPACE file, but it fails on trying to use the released version of `hermetic_cc_toolchain` as it does not have the new change.

@linzhp  Can you recommend the correct approach to fixing this? I don't see a way of doing this without breaking the API for `toolchains()` inside of `defs/toolchains.bzl`
